### PR TITLE
ENH: `isfinite` support for `datetime64` and `timedelta64`

### LINKED ISSUE #5610: Add `isfinite` support for `datetime64` and `timedelta64`
`numpy.isfinite` should support `datetime64` and `timedelta64` objects.  It should return `True` for any ordinary value, and `False` for `datetime64('NaT')` and `timedelta64('NaT')`.

As a side-effect, this will make `isclose` and `allclose` usable for `datetime64` and `timedelta64` objects.

```
In [97]: numpy.isfinite(numpy.datetime64(10**9, 's'))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-97-8555b18406e5> in <module>()
----> 1 numpy.isfinite(numpy.datetime64(10**9, 's'))

TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

### PR DESCRIPTION
**Pull request by debsankha**
This fixes #5610.

Works by adding a `PyUFunc_IsFiniteTypeResolver` for `isfinite` ufunc. It does what `PyUFunc_IsNaTTypeResolver` does for `datetime64` and `timedelta64` objects, and calls  `PyUFunc_DefaultTypeResolver` for all other dtypes.

### CODE REVIEW

**Comment by seberg** [comment]
I think quickly checking how to fix the isfinite version, we forgot that for histogram with automatic bins `np.linspace` would have to be fixed as well (not sure how straight forward that is). Although, I guess this probably still doesn't hurt, just does not quite fix the linked issue.

**Comment by debsankha** [comment]
@seberg : That's true. I looked around a bit, and found the following:

*  `np.linspace` is not the only thing stopping `np.histogram` from  working with `datetime64` and `timedelta64`. There are at least these issues:
    -  `np.histogram._get_outer_edges` tries to do addition/subtraction [between first_edge/last_edge and floats](https://github.com/numpy/numpy/blob/b30b8e24beacd6263f5978eb96733d57af536e89/numpy/lib/histograms.py#L320)  in order to avoid a division by zero. 
    - `np.histogram` [here tries to divide an int by `last_edge - first_edge`](https://github.com/numpy/numpy/blob/b30b8e24beacd6263f5978eb96733d57af536e89/numpy/lib/histograms.py#L813)
* `np.linspace` has only one problem with `timedelta64`: a [multiplication between two `timedelta64`](https://github.com/numpy/numpy/blob/b30b8e24beacd6263f5978eb96733d57af536e89/numpy/core/function_base.py#L161).
* `np.linspace` has at least this problems with `datetime64`:
    - [Multiplication with float](https://github.com/numpy/numpy/blob/b30b8e24beacd6263f5978eb96733d57af536e89/numpy/core/function_base.py#L135).

As per my novice understanding, both `linspace` and `histogram` needs special logic to deal with `datetime64` and `timedelta64`. And maybe some ufunc level chanhes might help, e.g. `multiply` between `timedelta64` and `float`.

**Comment by debsankha** [comment]
I have made an attempt to get `linspace ` to work with `datetime64` and `timedelta64`. Added (passing) tests as well. I had to pepper `linspace`  with datetime specific extra logic [in a few places](https://github.com/numpy/numpy/compare/master...debsankha:linspace-datetime#diff-12e00d917c1600a79611d57403cfbf70). 

Should I submit a PR?

File: numpy/core/code_generators/generate_umath.py [hunk 1]
```diff
@@ -844,8 +844,8 @@ def english_upper(s):
 'isfinite':
     Ufunc(1, 1, None,
           docstrings.get('numpy.core.umath.isfinite'),
-          None,
-          TD(nodatetime_or_obj, out='?'),
+          'PyUFunc_IsFiniteTypeResolver',
+          TD(nodatetime_or_obj + times, out='?'),
```
Comment (eric-wieser) [inline numpy/core/code_generators/generate_umath.py]: Shorter as just `noobj` (tested to be equivalent on my machine)

File: numpy/core/tests/test_datetime.py [hunk 2]
```diff
@@ -2208,6 +2208,27 @@ def test_isnat_error(self):
                 continue
             assert_raises(TypeError, np.isnat, np.zeros(10, t))
 
+    def test_isfinite(self):
+        assert_(not np.isfinite(np.datetime64('NaT', 'ms')))
+        assert_(not np.isfinite(np.datetime64('NaT', 'ns')))
+        assert_(np.isfinite(np.datetime64('2038-01-19T03:14:07')))
```
Comment (eric-wieser) [inline numpy/core/tests/test_datetime.py]: An amusing test :)

**Review by eric-wieser — APPROVED** [review]
Looks good to me - one tiny nit about the ufunc type letters.

**Comment by charris** [comment]
Needs a release note.

**Comment by charris** [comment]
@debsankha ping. All this needs is a release note (`doc/release/1.17.0-notes.rst`), and maybe use Eric's shorter type spec.

**Comment by debsankha** [comment]
OK, done.

**Comment by charris** [comment]
Thanks @debsankha .

**Pull request merged**